File: pset1/problems/credit/credit.c
#include <cs50.h>
#include <stdio.h>

// Prototypes
long get_number(void);
int check_luhn(long card_number);
int is_double_digit(int digit);

int main(void) {
    long card_number = get_number();
    int luhn_result = check_luhn(card_number);
    printf("Luhn is %i\n", luhn_result);
}

int is_double_digit(int digit) {
    if (digit * 2 > 9) {
        return ((digit * 2) % 10) + ((digit * 2) / 10);
    } else {
        return digit * 2;
    }
}

long get_number(void) {
    long card_number = get_long("Number: ");
    return card_number;
}

int check_luhn(long card_number) {
    /* Extract first 2 digits, validate checksum, count number of digits.
    Encode all information into an integer and return.

    Encoding rules:
    1. Validate checksum. If invalid return 0, else proceed.
    2. First 2 digits should stand for number of digits in the card. If
    number of digits is less than 13 or more than 16 return 0(invalid card).
    3. The last 2 digits of the integer should represent first 2 digits of
    the card
    */

    int first_two_digits;
    int count_digits = 0;

    // Sum of multiplied and not multiplied digits
    int sum_multiply = 0;
    int sum_not_multiply = 0;

    // Shrink card number
    while (card_number > 1000) {
        int last_digit = card_number % 10;
        sum_not_multiply += last_digit;
        card_number /= 10;
        count_digits++;

        int second_to_last = card_number % 10;
        // Check if product of the multiplied digit is more than 9. If so,
        // sum its digits.
        sum_multiply += is_double_digit(second_to_last);

        card_number /= 10;
        count_digits++;
    }

    // Process the remainder
    if (card_number < 100) {
        sum_not_multiply += card_number % 10;

        // Check if product of the multiplied digit is more than 9. If so,
        // sum its digits.
        sum_multiply += is_double_digit(card_number / 10);
        // Add 2 to count_digits
        count_digits += 2;

        // Add first digits
        first_two_digits = card_number;

    } else {
        sum_not_multiply += card_number % 10;
        card_number /= 10;

        // Check if product of the multiplied digit is more than 9. If so,
        // sum its digits.
        sum_multiply += is_double_digit(card_number % 10);
        sum_not_multiply += card_number / 10;
        // Add 3 to count_digits
        count_digits += 3;

        // Add first digits
        first_two_digits = card_number;
    }

    // Initialize total variable to store sum of digits
    int total = sum_not_multiply + sum_multiply;
    // Check conditions for checksum and for first digits
    if (total % 10 != 0) {
        return 0;
    }
    if (first_two_digits < 34 || first_two_digits > 55) {
        return 0;
    }

    count_digits *= 100;
    return (count_digits + first_two_digits);
}

File: pset6/problems/sentimental/credit.py
# TODO
def main():
    """Check credit card validity using Luhn's algorithm and first digits.

    This function serves as the entry point for the credit card validation
    program. It prompts the user to input a credit card number and then
    applies Luhn's algorithm to verify its validity. Additionally, it
    considers specific rules related to the first digits of the card to
    enhance the validation process.
    """
    number = get_number()
    valid, first_two_digits, count_digits = check_luhn(number)
    print_result(valid, first_two_digits, count_digits)


def get_number() -> int:
    """Prompt the user to input a positive integer.

    Returns
    -------
    int
        The positive integer entered by the user.
    """
    while True:
        try:
            number = int(input("Number: "))
        except ValueError:
            print("Invalid input.")
        else:
            if number > 0:
                return number
            else:
                print("Invalid input. Please enter a positive number")


def double_digit(digit: int) -> int:
    """Double the given digit and returns the result.

    If the doubled value is greater than 9, the function
    returns the sum of its digits; otherwise, it returns the
    doubled value as is.

    Parameters
    ----------
    digit : int
        The digit to be doubled.

    Returns
    -------
    int
        The result of doubling the given digit."""
    doubled = digit * 2
    return doubled % 10 + int(doubled / 10) if doubled > 9 else doubled


def check_luhn(number: int) -> tuple[bool, int, int]:
    """Check if a given number passes the Luhn's algorithm.

    Parameters
    ----------
    number : int
        The integer representing the credit card number.

    Returns
    -------
    Tuple[bool, int, int]
        A tuple containing:
        - A boolean indicating whether the number passed the Luhn's algorithm.
        - The first two digits of the credit card number.
        - The total count of digits in the credit card number.
    """
    card = number
    count = 0
    sum_multiply = 0
    sum_non_multiply = 0

    while card > 0:
        digit = card % 10
        if count % 2 == 0:
            sum_non_multiply += digit
        else:
            sum_multiply += double_digit(digit)
        card //= 10
        count += 1
    total = sum_multiply + sum_non_multiply
    # Return 0 if number didn't pass Luhn's algorithm or digit count less than
    # 13 or  more than 16.
    if total % 10 != 0 or count < 13 or count > 16:
        return (False, 0, 0)

    #  Extract first 2 digits from the card number
    first_two_digits = number // 10 ** (count - 2)
    if first_two_digits < 34 or first_two_digits > 55:
        return (False, 0, 0)
    return (True, first_two_digits, count)


def print_result(valid: bool, first_two_digits: int, count: int):
    """Print the result of credit card validation and, if valid, the card type.

    Parameters
    ----------
    valid : bool
        A boolean indicating whether the credit card number is valid.
    first_two_digits : int
        The first two digits of the credit card number.
    count : int
        The total count of digits in the credit card number.

    Returns
    -------
    None
        This function does not return any value. It prints the result directly.
    """
    if not valid or count not in {13, 15, 16}:
        print("INVALID")
        return
    # Visa card check
    elif 40 <= first_two_digits < 50 and count in {13, 16}:
        print("VISA")
        return
    # MasterCard check
    elif 50 < first_two_digits < 56 and count == 16:
        print("MASTERCARD")
        return
    # Amex card check
    elif first_two_digits in {34, 37} and count == 15:
        print("AMEX")
        return
    else:
        print("INVALID")
        return


if __name__ == "__main__":
    main()
